Support Bedrock tool caching

zoltan-ongithub opened the pull request:
Support cachePoint in tools.
Reference:
  https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html

## Title

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues

<!-- e.g. "Fixes #000" -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [X] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [X] I have added a screenshot of my new test passing locally 
- [X] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [X] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type
🆕 New Feature

![image](https://github.com/user-attachments/assets/d704696d-2391-4148-9ea1-bb4d23b46c8d)

## Changes

Support Bedrock tool caching. 
Documentation: https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html

Example usage:
```
tools = {...} # Tool definitions

tools.append({"cachePoint": {"type": "default"}}) # Enable tool caching
```

krrishdholakia commented on tests/llm_translation/test_bedrock_completion.py:
please add unit testing for your changes inside `tests/litellm/llms/bedrock/..`

krrishdholakia commented on litellm/litellm_core_utils/prompt_templates/factory.py:
```diff
@@ -3452,11 +3451,14 @@ def _bedrock_tools_pt(tools: List) -> List[BedrockToolBlock]:
 
     tool_block_list: List[BedrockToolBlock] = []
     for tool in tools:
+        if "cachePoint" in tool:
```
this is not the correct format - it should expect the anthropic spec, and translate accordingly - https://docs.litellm.ai/docs/completion/prompt_caching#anthropic-example

krrishdholakia replied:
reuse the similar logic as in _bedrock_converse_messages  - `litellm.AmazonConverseConfig()._get_cache_point_block` - _cache_point_block = (
                                litellm.AmazonConverseConfig()._get_cache_point_block(
                                    message_block=cast(
                                        OpenAIMessageContentListBlock, element
                                    ),
                                    block_type="content_block",
                                )
                            )
                            if _cache_point_block is not None:
                                _parts.append(_cache_point_block)

zoltan-ongithub committed:
Support Bedrock tool caching

Support cache_point in tools definition

Reference:

  https://docs.aws.amazon.com/bedrock/latest/userguide/prompt-caching.html
  https://docs.anthropic.com/en/docs/build-with-claude/prompt-caching

zoltan-ongithub commented:
@krrishdholakia, ready for next round.

zoltan-ongithub commented:
@krrishdholakia, ready for next round.
